fix(plugin-anthropic-proxy): SSE reverse-map leaks tokens straddling the tail-buffer flush cut

## Summary
createSseStream (src/proxy/sse-rewrite.ts) is the streaming reverse-map used by ProxyServer.handleRequest for every text/event-stream response. It retains the last SSE_TAIL_SIZE (64) chars in `pending` to reassemble patterns split across TCP chunks, then does `emit(reverseFn(flushable))` where `flushable = pending.slice(0, pending.length-64)`. The bug: it transforms the prefix IN ISOLATION. When a reverse-map token (e.g. a quoted CC tool name `"Write"`→`"write_file"`, `"KnowledgeSearch"`→`"recall"`, or any ELIZA_REVERSE_MAP/tool/prop-rename pattern) happens to start in the last ≤63 chars of `flushable` and end inside the retained tail, its head is emitted un-transformed (partial, no match) and its tail is transformed separately next round — so the full token is NEVER reverse-mapped and leaks through verbatim. On the reverse path this means the agent receives the raw upstream CC tool name (`"Write"`) instead of the eliza name (`"write_file"`), breaking tool-call recognition on the framework side, and can leak CC fingerprint strings into user-visible text. The failure is position-dependent (depends where the token lands relative to the internal 64-char cut in the accumulated buffer), so it is intermittent and hard to diagnose. The existing test only covers the cross-CHUNK split case, not the within-buffer straddle-the-cut case.

## Reproduction
```
Worktree probe test against src/proxy/sse-rewrite.ts (run with the repo's vitest):

  const emitted=[]; const stream=createSseStream(
    (t)=>t.replaceAll("ocplatform","elizaos"),
    (t)=>emitted.push(t), ()=>{});
  // single write, pattern at offset 32 of a 100-char payload so the internal
  // 64-char tail cut (sliceIdx=36) splits "ocplatform" into "ocpl"|"atform"
  stream.write(Buffer.from("x".repeat(32)+"ocplatform"+"y".repeat(58),"utf8"));
  stream.end();
  expect(emitted.join("")).not.toContain("ocplatform"); // FAILS

Observed: emitted output = "xxxx...ocplatformyyyy..." — the pattern is emitted verbatim, NOT replaced with "elizaos".
Expected: "ocplatform" reverse-mapped to "elizaos" (the pattern is fully present in the buffer at flush time; only the arbitrary internal cut splits it).
vitest output: `AssertionError: expected '...ocpla…' not to contain 'ocplatform'`. The analogous real leak is `"Write"`→`"write_file"` (ELIZA_TOOL_RENAMES) via reverseMap()/applyQuotedRenamesReverse in a streamed input_json_delta / content_block_start whose name token straddles the cut.
```

## Proposed fix
Bounded to src/proxy/sse-rewrite.ts + its test. Make the flush reverse-map a token that spans the cut before committing its head. Retain the last SSE_TAIL_SIZE raw chars for chunk-boundary reassembly (as today), but transform the flushable prefix WITH the retained tail as lookahead so a match beginning in flushable and ending in the tail is resolved before emit — i.e. reverse-map the whole `pending`, emit the mapped output minus the last SSE_TAIL_SIZE chars, and retain the raw tail for the next round (the eliza-side output tokens do not re-trigger the reverse keys — identity replacements and PascalCase→snake tool names are not quoted-PascalCase — so no double-mapping). Keep the existing surrogate-pair and StringDecoder handling. Test plan (>=10 lines, extends __tests__/sse-rewrite.test.ts): (1) single-write pattern positioned to straddle the internal 64-char cut → fully reverse-mapped; (2) parametrized sweep of the pattern across every offset in a >128-char buffer asserting no leak at any position; (3) real ELIZA_TOOL_RENAMES token `"Write"`→`"write_file"` straddling the cut inside a streamed input_json_delta; (4) regression: existing cross-chunk `ocp`+`latform` case still passes; (5) surrogate-pair split still intact; (6) idempotence: re-flushing already-mapped tail does not double-transform.

## Evidence plan
PR attaches: the failing probe (before) and passing suite (after) via `vitest run __tests__/sse-rewrite.test.ts`; the offset-sweep test output proving zero leaks across all boundary positions; a before/after streamed-response snippet showing `"name":"Write"` correctly reverse-mapped to `"name":"write_file"` when the token lands on the cut; and package typecheck/lint plus root `bun run verify`.

## Duplicate search
gh (agent-cortex, READ-ONLY) against elizaOS/eliza:
- `gh search issues --repo elizaOS/eliza "anthropic-proxy sse" / "sse-rewrite" / "createSseStream" / "reverse-map anthropic" / "plugin-anthropic-proxy stream"` → no relevant hits (only unrelated agent/core chat-streaming issues #17098/#17024/#16997).
- `gh search prs --repo elizaOS/eliza "sse-rewrite" / "createSseStream" / "anthropic-proxy sse"` → only unrelated SSE PRs in agent/core/plugin-codex-cli (#17416, #17397, #19262); nothing touching plugin-anthropic-proxy's tail-buffer.
No open or in-flight report for this createSseStream straddle-the-cut leak.

_Filed under a standing operator authorization for the autonomous audit pipeline, after local reproduction and duplicate search._

## Contribution provenance

- AI assistance: yes
- AI provider/model: anthropic / claude-opus-4-8
- Client / agent tooling: pi coding agent (eliza-swarm, autonomous)
- Attribution status: self-reported

<!-- eliza-computer-attribution:v1 {"provider":"anthropic","model":"claude-opus-4-8","client":"pi-coding-agent/eliza-swarm","skill_revision":"elizaOS/eliza@88ac1800e67fe2766881190d1af1a413844fa425:packages/skills/skills/contribute-to-eliza"} -->

--- a/plugins/plugin-anthropic-proxy/__tests__/sse-rewrite.test.ts
+++ b/plugins/plugin-anthropic-proxy/__tests__/sse-rewrite.test.ts
@@ -5,6 +5,8 @@
  */
 
 import { describe, expect, it } from "vitest";
+import { ELIZA_TOOL_RENAMES } from "../src/proxy/eliza-fingerprint.js";
+import { reverseMap } from "../src/proxy/reverse-map.js";
 import { createSseStream } from "../src/proxy/sse-rewrite.js";
 
 describe("createSseStream", () => {
@@ -45,6 +47,129 @@ describe("createSseStream", () => {
     const joined = emitted.join("");
     expect(joined).toContain("中文 🚀");
     expect(joined).not.toContain("\uFFFD");
+  });
+
+  it("reverse-maps a token that straddles the internal 64-char tail cut", () => {
+    // Single write: the pattern lands at offset 32 of a 100-char payload, so
+    // the internal tail cut (mapped.length - 64) falls inside "ocplatform".
+    // Transforming the flushable prefix in isolation left the head un-mapped
+    // and leaked the pattern verbatim (issue #21257).
+    const emitted: string[] = [];
+    const stream = createSseStream(
+      (text) => text.replaceAll("ocplatform", "elizaos"),
+      (text) => emitted.push(text),
+      () => undefined
+    );
+
+    stream.write(Buffer.from(`${"x".repeat(32)}ocplatform${"y".repeat(58)}`, "utf8"));
+    stream.end();
+
+    const joined = emitted.join("");
+    expect(joined).not.toContain("ocplatform");
+    expect(joined).toContain("elizaos");
+    expect(joined).toBe(`${"x".repeat(32)}elizaos${"y".repeat(58)}`);
+  });
+
+  it("never leaks the pattern at any offset across a >128-char buffer", () => {
+    // Sweep the pattern across every start offset so no boundary position
+    // (relative to the internal cut) can leak. Adversarial coverage of the
+    // position-dependent failure.
+    const pattern = "ocplatform";
+    const total = 160;
+    for (let offset = 0; offset + pattern.length <= total; offset++) {
+      const before = "x".repeat(offset);
+      const after = "y".repeat(total - offset - pattern.length);
+      const payload = before + pattern + after;
+      const emitted: string[] = [];
+      const stream = createSseStream(
+        (text) => text.replaceAll(pattern, "elizaos"),
+        (text) => emitted.push(text),
+        () => undefined
+      );
+      stream.write(Buffer.from(payload, "utf8"));
+      stream.end();
+      const joined = emitted.join("");
+      expect(joined, `offset ${offset}`).not.toContain(pattern);
+      expect(joined, `offset ${offset}`).toBe(`${before}elizaos${after}`);
+    }
+  });
+
+  it('reverse-maps a real "Write" tool name that straddles the cut in an input_json_delta', () => {
+    // The analogous production leak: an SSE input_json_delta / content_block_start
+    // whose CC tool name ("Write") lands on the internal cut must still reverse
+    // to the eliza name ("write_file") so tool-call recognition works.
+    const config = {
+      toolRenames: ELIZA_TOOL_RENAMES,
+      propRenames: [] as ReadonlyArray<readonly [string, string]>,
+      reverseMap: [] as ReadonlyArray<readonly [string, string]>,
+    };
+    const reverse = (text: string) => reverseMap(text, config);
+    const suffix = `"input":{}}${"z".repeat(60)}`;
+    // Position "Write" so the 64-char tail cut splits the quoted token.
+    const prefix = `event: content_block_start\ndata: {"type":"tool_use",${"a".repeat(10)}`;
+    const payload = `${prefix}"name":"Write",${suffix}`;
+
+    const emitted: string[] = [];
+    const stream = createSseStream(
+      reverse,
+      (text) => emitted.push(text),
+      () => undefined
+    );
+    stream.write(Buffer.from(payload, "utf8"));
+    stream.end();
+
+    const joined = emitted.join("");
+    expect(joined).toContain('"name":"write_file"');
+    expect(joined).not.toContain('"name":"Write"');
+  });
+
+  it('reverse-maps "Write" no matter where it lands relative to the cut', () => {
+    const config = {
+      toolRenames: ELIZA_TOOL_RENAMES,
+      propRenames: [] as ReadonlyArray<readonly [string, string]>,
+      reverseMap: [] as ReadonlyArray<readonly [string, string]>,
+    };
+    const reverse = (text: string) => reverseMap(text, config);
+    const token = '"name":"Write"';
+    for (let offset = 0; offset <= 140; offset += 1) {
+      const payload = `${"a".repeat(offset)}${token}${"b".repeat(140 - offset)}`;
+      const emitted: string[] = [];
+      const stream = createSseStream(
+        reverse,
+        (text) => emitted.push(text),
+        () => undefined
+      );
+      stream.write(Buffer.from(payload, "utf8"));
+      stream.end();
+      const joined = emitted.join("");
+      expect(joined, `offset ${offset}`).toContain('"name":"write_file"');
+      expect(joined, `offset ${offset}`).not.toContain('"name":"Write"');
+    }
+  });
+
+  it("does not double-transform an already-mapped retained tail (idempotence)", () => {
+    const config = {
+      toolRenames: ELIZA_TOOL_RENAMES,
+      propRenames: [] as ReadonlyArray<readonly [string, string]>,
+      reverseMap: [] as ReadonlyArray<readonly [string, string]>,
+    };
+    const reverse = (text: string) => reverseMap(text, config);
+    // Drip the token in one-byte chunks so its mapped form sits in the retained
+    // tail across many rounds; the eliza output token "write_file" must not be
+    // re-triggered by any reverse key.
+    const payload = `${"c".repeat(80)}"name":"Write"${"d".repeat(80)}`;
+    const emitted: string[] = [];
+    const stream = createSseStream(
+      reverse,
+      (text) => emitted.push(text),
+      () => undefined
+    );
+    for (const byte of Buffer.from(payload, "utf8")) {
+      stream.write(Buffer.from([byte]));
+    }
+    stream.end();
+    const joined = emitted.join("");
+    expect(joined).toBe(`${"c".repeat(80)}"name":"write_file"${"d".repeat(80)}`);
   });
 
   it("calls finish even for empty streams", () => {

--- a/plugins/plugin-anthropic-proxy/src/proxy/sse-rewrite.ts
+++ b/plugins/plugin-anthropic-proxy/src/proxy/sse-rewrite.ts
@@ -5,6 +5,25 @@
  * Without this, "ocplatform" can split as "ocp"+"latform" and leak through.
  * TAIL_SIZE >= longest reverseMap pattern.
  *
+ * The reverse map is applied to the WHOLE accumulated buffer before any bytes
+ * leave, then only the mapped output beyond the retained tail is emitted. This
+ * matters because a token can straddle the internal tail cut: it starts in the
+ * flushable prefix and ends inside the retained tail. Transforming the prefix
+ * in isolation would emit its head un-mapped and map its tail separately next
+ * round, so the full token would never reverse-map and would leak verbatim
+ * (e.g. `"Write"` reaching the framework instead of `"write_file"`). Mapping
+ * the whole buffer first resolves any such straddling token before its head is
+ * committed.
+ *
+ * The retained tail is the already-mapped suffix, not raw bytes. reverseFn must
+ * be idempotent — its eliza-side output tokens (snake_case tool names, identity
+ * replacements) do not re-trigger the quoted-PascalCase reverse keys — so a
+ * settled tail is never double-transformed on the next round, while an
+ * incomplete key at a real chunk boundary (kept literal because it did not
+ * match) still completes and maps once its continuation arrives. TAIL_SIZE >=
+ * longest quoted/escaped reverse key guarantees such an incomplete key is
+ * always fully within the retained tail.
+ *
  * Also uses StringDecoder to buffer partial UTF-8 sequences across TCP
  * chunks. chunk.toString() would emit U+FFFD whenever a multi-byte char
  * (中文, emoji, etc.) lands on a chunk boundary.
@@ -35,20 +54,26 @@ export function createSseStream(
   return {
     write(chunk: Buffer): void {
       pending += decoder.write(chunk);
-      if (pending.length > SSE_TAIL_SIZE) {
-        let sliceIdx = pending.length - SSE_TAIL_SIZE;
+      // Map the whole buffer so a token straddling the internal tail cut is
+      // resolved before its head is emitted; retain the mapped tail (not raw)
+      // for the next round.
+      const mapped = reverseFn(pending);
+      if (mapped.length > SSE_TAIL_SIZE) {
+        let sliceIdx = mapped.length - SSE_TAIL_SIZE;
         // Don't cut between a UTF-16 surrogate pair
-        const prev = pending.charCodeAt(sliceIdx - 1);
+        const prev = mapped.charCodeAt(sliceIdx - 1);
         if (prev >= 0xd800 && prev <= 0xdbff) sliceIdx -= 1;
-        const flushable = pending.slice(0, sliceIdx);
-        pending = pending.slice(sliceIdx);
-        emit(reverseFn(flushable));
+        emit(mapped.slice(0, sliceIdx));
+        pending = mapped.slice(sliceIdx);
+      } else {
+        pending = mapped;
       }
     },
     end(): void {
       pending += decoder.end();
-      if (pending.length > 0) {
-        emit(reverseFn(pending));
+      const mapped = reverseFn(pending);
+      if (mapped.length > 0) {
+        emit(mapped);
       }
       finish();
     },

